File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        from collections import deque
        q = deque([root])
        results = []
        while q:
            level = []
            level_size = len(q)
            for _ in range(level_size):

                node = q.popleft()
                if node is None:
                    return results
                if node.left is not None:
                    q.append(node.left)
                if node.right is not None:
                    q.append(node.right)
                level.append(node.val)
            results.append(level)
        return results

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-5.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(p->val < root->val && q->val < root->val){
            root = root->left;
            return lowestCommonAncestor(root, p, q);
        }
        else if(p->val > root->val && q->val > root->val) {
            root = root -> right;
            return lowestCommonAncestor(root, p, q);
        }
        return root;
    }
};
